src/java.base/share/classes/java/security/SecureRandom.java

```diff
@@ -267,23 +275,34 @@ public SecureRandom(byte[] seed) {
     private void getDefaultPRNG(boolean setSeed, byte[] seed) {
         Service prngService = null;
         String prngAlgorithm = null;
-        for (Provider p : Providers.getProviderList().providers()) {
-            // SUN provider uses the SunEntries.DEF_SECURE_RANDOM_ALGO
-            // as the default SecureRandom algorithm; for other providers,
-            // Provider.getDefaultSecureRandom() will use the 1st
-            // registered SecureRandom algorithm
-            if (p.getName().equals("SUN")) {
-                prngAlgorithm = SunEntries.DEF_SECURE_RANDOM_ALGO;
+
+        // If in FIPS mode, use the SecureRandom from the FIPS provider
+        if (FIPSConfigurator.enableFIPS()) {
+            Provider p = Security.getProvider("SunPKCS11-NSS-FIPS");
+            prngAlgorithm = "PKCS11";
+            if (p != null) {
                 prngService = p.getService("SecureRandom", prngAlgorithm);
```

Comment:
Should this not fail in some way if `prngService == null` if FIPS mode is requested?

Reply:
> 

I added two exceptions, one for not finding the `SunPKCS11-NSS-FIPS` provider, the other one for not getting the `PKCS11` correctly.

Reply:
Suggest inverting the if statements and adjusting the exception messages (which typically don't end with a period):
```
            if (p == null) {
                throw new RuntimeException("could not find SunPKCS11-NSS-FIPS provider for FIPS mode");
            }
            prngService = p.getService("SecureRandom", prngAlgorithm);
            if (prngService == null) {
                throw new RuntimeException("could not find SecureRandom implementation from SunPKCS11-NSS-FIPS");
            }
```

Reply:
inverted